Add Core Built-In Tool Capabilities to the Upsonic Runtime

samitugal opened the pull request:
# Overview 

This PR implements a complete tool system for the Upsonic project, including a flexible decorator-based architecture, three production-ready tool classes, and a comprehensive test suite. The system provides a clean, extensible way to create and manage AI agent tools with manual control over tool assignment.

# Motivation
Nowadays, most modern frameworks provide their own set of built-in tools for users. I noticed that the Upsonic Framework currently does not include such a tool library. While I’m very satisfied using Upsonic in my work, this limitation has caused some difficulties for me.

That’s why I wanted to contribute this improvement. For now, I’ve added built-in tools for yfinance, duckducksearch, and Tavily Web Search. More tools can be added in the future if needed.

# Core System Architecture for Tools
* @tool decorator: Marks methods as tools without auto-registration
* get_tools_from_instance(): Extracts marked methods from class instances
* functions() method: Standardized API for tool retrieval
* Manual tool assignment: Full control over which tools are assigned to tasks

```python
tool_instance = SomeToolClass()
tools = tool_instance.functions()
task = Task("description", tools=tools1 + tools2)
```

>While defining the tools for a task, I don’t really like the way tools are combined using a + operator. I found a solution for this: the tool information should be flattened. However, since this logic is inside the agent definitions, I deliberately left it as-is for now.

# Tool Classes Implemented
## 1. TavilySearchTool (src/upsonic/tools/websearch/tavily.py)
* Methods: search_detailed(), search_context()
* Features:
  * JSON and Markdown output formats
  * Token limit processing
  * Configurable search depth ("basic", "advanced")
  * Environment variable and explicit API key support
  * Custom max_results and include_answer options
## 2. DuckDuckGoSearchTool (src/upsonic/tools/websearch/duckduckgo.py)
* Methods: search()
* Features:
  * Simple web search with customizable result limits
  * Special character handling
  * Consistent API with other tools
  * Added missing functions() method
## 3. YFinanceTools (src/upsonic/tools/finance/yfinance.py)
* Methods:
  * get_current_stock_price()
  * get_company_info()
  * get_analyst_recommendations()
  * get_company_news()
  * get_stock_fundamentals()
  * get_income_statements()
  * get_key_financial_ratios()
  * get_historical_stock_prices()
  * get_technical_indicators()
* Features: Comprehensive financial data access with error handling

# Directory Structure

```
src/upsonic/tools/
├── __init__.py                     # Main tool exports
├── decorators/
│   └── tool_decorator.py          # @tool decorator system
├── websearch/
│   ├── __init__.py                # Web search tool exports
│   ├── tavily.py                  # Tavily search implementation
│   └── duckduckgo.py              # DuckDuckGo search implementation
├── finance/
│   ├── __init__.py                # Finance tool exports
│   └── yfinance.py                # Yahoo Finance implementation
├── tool_registry.py              # Tool registry system
└── tool.py                       # Base tool interface
```
# Tool System Implementation
## Creating a New Tool Class
```python
from upsonic.tools.decorators.tool_decorator import tool, get_tools_from_instance

class MyCustomTool:
    def __init__(self, api_key: str = None):
        self.api_key = api_key
    
    @tool(description="Perform a custom search operation")
    def search(self, query: str, limit: int = 10) -> str:
        """Search for information using custom logic"""
        # Your implementation here
        return f"Search results for: {query}"
    
    @tool(description="Process data with custom algorithm")
    def process_data(self, data: str, format: str = "json") -> str:
        """Process input data and return formatted results"""
        # Your implementation here
        return f"Processed {data} in {format} format"
    
    def helper_method(self):
        """This method won't be exposed as a tool (no @tool decorator)"""
        return "Helper function"
    
    def functions(self):
        """Return the list of tool functions to be used by the agent."""
        return get_tools_from_instance(self)
```

## Using Tools with Agents
```python
from upsonic import Agent, Task

# Initialize tool instances
custom_tool = MyCustomTool(api_key="your_key")
finance_tool = YFinanceTools()
search_tool = TavilySearchTool(api_key="tavily_key")

# Extract tool methods
custom_tools = custom_tool.functions()
finance_tools = finance_tool.functions()
search_tools = search_tool.functions()

# Combine tools as needed
all_tools = custom_tools + finance_tools + search_tools

# Create task with specific tools
task = Task(
    description="Find information about AAPL stock and general AI trends",
    tools=all_tools  # Manual tool assignment
)

# Execute with agent
agent = Agent(name="Research Agent")
result = agent.do(task)
```

Best Practices
* Always include descriptions: The @tool decorator requires a clear description
* Use type hints: Helps with validation and documentation
* Return strings: Agents work best with string responses
* Handle errors gracefully: Return error messages as strings instead of raising exceptions
* Keep methods focused: Each tool method should do one specific thing
* Include validation: Validate inputs and return meaningful error messages
* Use the functions() method: Standardize tool extraction across all classes

# Tool System Test Suite
Comprehensive test suite for the Upsonic tool system with 62 test cases achieving 100% success rate. The tests ensure reliability, prevent regressions, and provide usage examples for all tool functionality.

## Test Structure
tests/tools/
```
├── test_tavily.py           # 26 tests - Tavily search tool
├── test_duckduckgo.py       # 12 tests - DuckDuckGo search tool  
├── test_yfinance.py         # 21 tests - Yahoo Finance tools
├── test_tool_decorator.py   # 15 tests - @tool decorator system
├── run_all_tests.py         # Test runner with reporting
└── __init__.py             # Package initialization
```

## Test Coverage

* TavilySearchTool (26 tests)
  * API key initialization (env vars + explicit)
  * Search methods (search_detailed, search_context)
  * Output formats (JSON, Markdown)
  * Token limiting and error handling
  * Custom parameters and edge cases
*DuckDuckGoSearchTool (12 tests)
  * Basic search functionality
  * Result limits and special characters
  * Empty results and exception handling
  * Method attributes validation
* YFinanceTools (21 tests)
  * All 9 financial methods tested
  * Success scenarios and error cases
  * Data format validation
  * Mock API integration
* Tool Decorator System (15 tests)
  * @tool decorator functionality
  * get_tools_from_instance() behavior
  * Method binding and inheritance
  * Complex signatures and error handling

## Running Tests

```bash
cd tests/tools
python run_all_tests.py
```

# Related Issue  
Closes #384

samitugal commented:
Hi @onuratakan,  
It seems that the AWS CodeBuild pipeline is currently broken — I noticed that other recent PRs are also failing for the same reason.  Could you please check my PR once the build system is fixed? ,
 
All tests pass locally on my side and the implementation is ready for review.  
Thanks

onuratakan commented:
Hi, thanks and one more think,

The .functions() is cool but users shouldn't  struggle with that. When we get the tools in `agent_tool_register` we need to analyze the tool format and if its an cls we should try to call it .functions() automatically.

https://github.com/Upsonic/Upsonic/blob/master/src/upsonic/utils/direct_llm_call/agent_tool_register.py#L1

With that users no longer need to call and creating a list to provide.

Can you add that?




Current:
```python
from upsonic import Agent, Task

# Initialize tool instances
custom_tool = MyCustomTool(api_key="your_key")
finance_tool = YFinanceTools()
search_tool = TavilySearchTool(api_key="tavily_key")

# Extract tool methods
custom_tools = custom_tool.functions()
finance_tools = finance_tool.functions()
search_tools = search_tool.functions()

# Combine tools as needed
all_tools = custom_tools + finance_tools + search_tools

# Create task with specific tools
task = Task(
    description="Find information about AAPL stock and general AI trends",
    tools=all_tools  # Manual tool assignment
)

# Execute with agent
agent = Agent(name="Research Agent")
result = agent.do(task)
```


We need to turn that to:
```python
from upsonic import Agent, Task

# Initialize tool instances
custom_tool = MyCustomTool(api_key="your_key")
finance_tool = YFinanceTools()
search_tool = TavilySearchTool(api_key="tavily_key")


# Create task with specific tools
task = Task(
    description="Find information about AAPL stock and general AI trends",
    tools=[custom_tools, finance_tools, search_tools]  # Manual tool assignment
)

# Execute with agent
agent = Agent(name="Research Agent")
result = agent.do(task)
```

samitugal commented:
Hi

I’ve completed all the requested improvements — everything now works as expected.

```python
task = Task(
    "Search for information about 'artificial intelligence' and get the current stock price for AAPL also how many Chai Tea Latte are in the warehouse?",
    tools=[tavily_tool, yfinance_tools, custom_tool],
)

agent = Agent(name="Search Agent")
print(agent.print_do(task))
```

I also added a validation to the Task class so that users aren’t forced to wrap a single tool in a list. If a single tool is provided, the validator automatically converts it into a list.

```python
class Task(BaseModel):
    description: str
    images: Optional[List[str]] = None
    tools: list[Any] = []
    response_format: Union[Type[ObjectResponse], Type[BaseModel], type[str], None] = str
    response_lang: Optional[str] = "en"
    _response: Any = None
    context: Any = []
    price_id_: Optional[str] = None
    task_id_: Optional[str] = None
    not_main_task: bool = False
    start_time: Optional[int] = None
    end_time: Optional[int] = None
    agent: Optional[Any] = None
    _tool_calls: List[Dict[str, Any]] = []

    @field_validator("tools", mode="before")
    @classmethod
    def validate_tools_format(cls, v):
        """Validate and normalize tools to list format"""
        if v is None:
            return []

        if not isinstance(v, list):
            return [v]

        return v
```

This means you can now define a task like this as well:

```python
task = Task(
    "Search for information about 'artificial intelligence'",
    tools=tavily_tool,
)
```

In addition, I noticed that the response_lang field was duplicated in the Task class fields, so I cleaned that up too.

I hope you like it

samitugal commented:
## Add Toolkit Abstract Base Class for Tool Standardization

### Overview
This PR introduces a new `Toolkit` abstract base class to standardize all tool implementations, improves tool validation in the `Task` class, and enhances type safety across the codebase.

### What Changed

#### 1. **New Base Class: `Toolkit`**
- Created `src/upsonic/tools/base/toolkit.py` with ABC implementation
- All tools now inherit from this base class
- Provides common interface for tool discovery and validation

#### 2. **Automatic Tool Discovery**
- Added `functions()` method that automatically finds `@tool` decorated methods
- No need to manually implement this method in each tool class
- Uses existing `get_tools_from_instance()` utility

#### 3. **Tool Validation System**
- Added `__control__()` method for tool health checks
- Validates API keys, connections, and readiness before use
- Can be overridden by each tool for custom validation logic

#### 4. **Enhanced Task Validation**
- **Fixed incomplete `validate_tools()` method** - Previously it called `__control__()` but ignored the result
- **Added proper error handling** with try-catch blocks and meaningful error messages
- **Improved field validation** in `validate_tools_format()` to ensure only `Toolkit` instances are accepted
- **Runtime type checking** with helpful developer feedback

#### 5. **Type Safety Improvements**
- Added field validator in `Task` class to ensure only `Toolkit` instances are accepted
- Clear error messages when invalid tools are provided
- Runtime type checking with helpful developer feedback

#### 6. **Updated Existing Tools**
- `TavilySearchTool` now inherits from `Toolkit`
- `YFinanceTools` now inherits from `Toolkit`
- Removed duplicate `functions()` methods from existing tools

### Code Example

**Before:**
```python
class CustomTool:
    def functions(self):
        return get_tools_from_instance(self)
    
    @tool(description="My tool")
    def my_method(self):
        return "result"

# Task validation was incomplete
def validate_tools(self):
    control_result = tool.__control__()  # Result was ignored!
```

**After:**
```python
class CustomTool(Toolkit):
    def get_description(self) -> str:
        return "Description of what this tool does"
    
    def __control__(self) -> bool:
        # Custom validation logic
        return self.api_key is not None
    
    @tool(description="My tool")
    def my_method(self):
        return "result"

# Task validation now properly handles results
def validate_tools(self):
    try:
        control_result = tool.__control__()
        if not control_result:
            raise ValueError(f"Tool validation failed for {tool.__class__.__name__}")
    except Exception as e:
        raise ValueError(f"Tool validation error: {str(e)}")
```

### Benefits

✅ **Standardized Interface** - All tools follow the same pattern  
✅ **Type Safety** - Runtime validation ensures only valid tools are used  
✅ **Better Error Messages** - Clear feedback when tools are incorrectly implemented  
✅ **Automatic Discovery** - No boilerplate code for tool registration  
✅ **Health Checks** - Tools can validate their readiness before use  
✅ **Robust Validation** - Task class now properly validates and handles tool errors  
✅ **Clean Code** - Removes code duplication across tool implementations  
✅ **Early Error Detection** - Issues are caught before task execution begins  

### Breaking Changes
- Tool classes must now inherit from `Toolkit`
- Must implement abstract `get_description()` method
- Invalid tools will raise validation errors at runtime
- Tool validation errors are now properly surfaced (previously ignored)

### Task Validation Improvements
- **Fixed Bug**: `validate_tools()` method now actually validates `__control__()` results
- **Error Handling**: Proper exception handling with descriptive error messages
- **Type Safety**: Field validator ensures only `Toolkit` instances are accepted
- **Developer Experience**: Clear error messages indicate exactly what's wrong

### Migration Guide
For existing tools, change:
```python
class YourTool:
    def functions(self):
        return get_tools_from_instance(self)
```

To:
```python
class YourTool(Toolkit):
    def get_description(self) -> str:
        return "Your tool description"
    
    def __control__(self) -> bool:
        # Optional: Add custom validation
        return True
```

### Testing
- Added validation tests to ensure type safety works correctly
- All existing tools updated and tested
- Tool health check system validated
- Task validation improvements tested with both valid and invalid tools

This change makes the tool system more robust and developer-friendly while fixing existing validation issues and maintaining backward compatibility for tool functionality.
